### ISSUE TITLE
Unable to bind mount on certain hard drives

### ISSUE THREAD

**Issue by matthewbauer** [NixOS/nixpkgs#67465]
**Describe the bug**
This has been reported in two discourse threads:

- https://discourse.nixos.org/t/nixos-install-unable-to-bind-mount/3786
- https://discourse.nixos.org/t/unable-to-nixos-install-from-live-cd/2533/6

The error appears when trying to install NixOS on a hard drive. It appears some hardware issue might be going on, along with maybe a Nix issue:

```
error: while setting up the build environment: unable to bind mount '/mnt/nix/store/85d8y2vybq07al01j7zv518yf3fa9mwq-append-initrd-secrets.drv.chroot': Permission denied
```

**Comment by alexeymuranov** [comment]
I am trying to install NixOS on a USB flash drive in BIOS mode (to boot from it some machine with Windows), and I am getting
```
error: while setting up the build environment: unable to bind mount '/home/alexey/.mnt/nix/store/0vq9ja754cj161nwq8b8945p5b94fhnj-NetworkManager.conf.drv.chroot': Permission denied
```

If i disable `networkmanager` in `configuration.nix`, i get the same error for `append-initrd-secrets`.

**Comment by EliotBerriot** [comment]
facing the same issue today. Trying to install NixOS in a separate partition on my laptop which already runs PureOS on top of Luks.

Trying a live CD install and installing from whithing PureOS itself raises: 

```
error: while setting up the build environment: unable to bind mount '/mnt/nix/store/7gr4d45v9mhllcpf0grh73nf75br639c-Xresources-Xft.drv.chroot': Permission denied
```

**Comment by EliotBerriot** [comment]
I've tried to do the install on a non-encrypted partition, with the same result (so at least we know it's not related to LUKS/disk encryption)

**Comment by alyssais** [comment]
A couple of things you could try:

- run `nixos-install` under `strace -f` and post the output. It will likely be quite big but should compress well. (One of the discourse posts has strace output but not with `-ﬂ so we can’t see what children are doing.)
- Try installing NixOS unstable and see if that changes anything.

**Comment by EliotBerriot** [comment]
Thanks @alyssais, I'll give it a go, probably tomorrow!

**Comment by EliotBerriot** [comment]
So, I've tried using the unstable channel, with the same result.

@alyssais this is the output of my latest attempt, run with `sudo PATH="$PATH" NIX_PATH="$NIX_PATH" strace -f -o /tmp/strace_output.txt `which nixos-install` --root /mnt`:

[strace_output.txt](https://github.com/NixOS/nixpkgs/files/4169726/strace_output.txt)

**Comment by TinHead** [comment]
Hi, 
Getting the same on a Surface Laptop 3 running Manjaro. Trying to install NixOS on another partition.

SSD: 01:00.0 Non-Volatile memory controller: SK hynix BC501 NVMe Solid State Drive 512GB

[strace_output.txt](https://github.com/NixOS/nixpkgs/files/4540644/strace_output.txt)

Cheers!

**Comment by matthewbauer** [comment]
Could it be some mount flag is messing things up? Maybe try:

```
mount | grep /mnt
```

I wonder if we are hitting some inodes limit?

**Comment by TinHead** [comment]
Nothing special there.

`/dev/nvme0n1p7 on /mnt type ext4 (rw,relatime)`

Inodes seem to be fine:
```sudo df -ih /
Filesystem     Inodes IUsed IFree IUse% Mounted on
/dev/nvme0n1p5   3,2M  550K  2,7M   17% /
LixSurface% sudo df -ih /mnt
Filesystem     Inodes IUsed IFree IUse% Mounted on
/dev/nvme0n1p7    13M   53K   13M    1% /mnt
```

**Comment by alexeymuranov** [comment]
I am consistently getting this error with `NetworkManager.conf.drv.chroot`, this time trying NixOS 20.03.

So this is not purely hardware related.

**Comment by jtrakk** [comment]
I also get it with `/mnt/dev/nix/store/gispa2laayl2bh2v5dsk3pmcsp943mgc-NetworkManager.conf.drv.chroot` on NixOS 20.09pre226148.0f5ce2fac0c.

**Comment by energizah** [comment]
I can reproduce this issue with a minimal `configuration.nix`. I'm including some debugging info in [this repo](https://github.com/energizah/nixos-bind-mount-error).

**Comment by energizah** [comment]
The parent directory of the installation target had insufficient permissions. `chmod o+rx` on that directory resolved the error.

Thanks to @euank for solving this.

**Comment by matthewbauer** [comment]
> The parent directory of the installation target had insufficient permissions. `chmod o+rx` on that directory resolved the error.
> 
> Thanks to @euank for solving this.

Is there a change we can make to nixos-install that can make this work out of the box? I’m a little confused still that this only happens in certain cases.

**Comment by euank** [comment]
Yeah, I can explain the cases that this happens in, and also what options we have to improve things here.

First, the easiest way to reproduce this is to just run `chmod o-rx /mnt` and try to do an install that will require building a derivation not in either your local cache or the remote cache.

That last bit is why it only sometimes happens. If `nix` downloads a file from a cache into a `/mnt/root/nix/store` with bad `/mnt` permissions, all is fine there. Even though `/mnt` forbids 'other' access, `nix` is running as root and can ignore those permissions entirely (root has CAP_FOWNER more specifically).

However, if nix has to build a derivation, it has to spin up a builder... And builders run in a user namespace with uids and gids mapped and significantly less permissions. That builder is what's spitting out that error.

We've specifically seen this with the [`NetworkManager.conf` derivation](https://github.com/NixOS/nixpkgs/blob/9d0c3ffe6783d59b427d018e8341e0084737fde9/nixos/modules/services/networking/networkmanager.nix#L24-L52), because, well, that's not really something that can be copied from the cache (requires mac address, etc) and so is getting instantiated each time.

Hopefully the above makes sense. Basically, you don't need a reasonable tree of permissions for copying stuff into the nix store, but you do for nix to spawn a builder that then operates on that store.

So, what are our options to fix this?
1. nix could do a better job of recognizing `EACCES` and try to present a useful error message, possibly walking up the filesystem tree itself to do so.
Unfortunately, that seems kinda tricky to do correctly. The kernel gives us back EACCES, and it's not like we can actually derive why reliably (i.e. we don't know if it's selinux, or filesystem permissions or what).
2. nix-install could add additional validation.
I implemented this over here:  #90431
Basically, it can walk the tree itself and verify permissions seem right.

**Comment by alexeymuranov** [comment]
Hm, in my case the mount point was `~.mnt` with permissions `drwxr-xr-x`.

How to explain that re-partitionning a flash drive seemed to have helped once?

**Comment by euank** [comment]
> Hm, in my case the mount point was ~/.mnt with permissions drwxr-xr-x. The parent directories, including /home, have the same permissions (with different ownership).


That's unusual; normally `/home/$USER` will have permissions of `700`, which would be consistent with my explanation.

You could try, for example, `sudo -u nixbld1 /bin/sh` and then `ls /home/$user/.mnt/nix/store` to see whether it's as straight-forward a permissions problem as I describe above, where the builder just doesn't have access to that filesystem path.

> How to explain that re-partitionning a flash drive seemed to have helped once? (Or maybe not...)

It's very possible there are multiple issues here as well with the same error.
We ended up isolating this one with `perf ftrace -G do_mount -- <command>` to get a little more information where in the kernel it was getting EACCES, so if you have a different cause than permissions, the output of that would be helpful for understanding it better!

**Comment by TinHead** [comment]
chmod o+rx fixed it for me too!

**Comment by alexeymuranov** [comment]
@euank, sorry, i do not know how i was looking. In fact, my `~/.mnt` directory had "wrong" permissions. I had `~/.mnt` and `~/.mnt2`, and `~/.mnt2`'s permissions were "right", but `~/.mnt`'s permissions were "wrong". When using `drwxr-xr-x` permissions, installation goes through fine. Sorry for the fuss.

So, maybe, when i thought that re-partitionning the flash drive fixed the issue, it just happened that i mounted it under `~/.mnt2` instead of under `~/.mnt`.

**Comment by matthewbauer** [comment]
Closing with https://github.com/NixOS/nixpkgs/pull/90431 merged. Now nixos-install will give you a warning when you accidentally do this.
